### ISSUE TITLE
string_span's string_length and strnlen fail for arm-none-eabi

### ISSUE THREAD

**Issue by Josaphat** [Microsoft/GSL#391]
Including the use of `strnlen` is technically un-portable because it is not specified by the C or C++ standards, so using `-std=c++11` with gcc *should* fail IMO because it's providing non-standard functionality. But really I don't have an issue with the GSL using `strnlen` beyond the fact that it doesn't work on my platform (arm-none-eabi-gcc). The solution that was submitted in #334 is almost there but there are two short-comings:

1. Although my platform is not cygwin, `strnlen` is not declared in `string.h` if `std=c++..` is specified. arm-none-eabi-gcc doesn't allow me to use `-isystem` for c++ files because it implicitly wraps system includes with `extern "C" { }`, so gsl must be in my regular include path. I compile with `-Werror` so the implicit function declaration of `strnlen` causes my builds to fail.
2. `string_length` fails to compile cleanly because of an implicit narrowing conversion from `std::size_t` to `int` when `strnlen` is not used:
```
    span<const char> str_span{str, n}; // n is declared as std::size_t
```

To fix (1), is it reasonable to simply check if the symbol `strnlen` is defined?
```
#ifdef strnlen
#define GSL_PLATFORM_HAS_STRNLEN
#endif // strnlen
```

If this isn't a good solution then I would propose not defining `GSL_PLATFORM_HAS_STRNLEN` directly in the code at all and assuming that the platform does not have strnlen. If users of GSL require `string_length` to use `strnlen`, they could define `GSL_PLATFORM_HAS_STRNLEN` themselves.

For (2) I would simply add a checked narrowing conversion: `gsl::narrow<int>(n)`.

I would happily submit a PR which includes these changes.

**Comment by MikeGitb** [comment]
Radical suggestion: If the GSL has to provide it's own implementation of strlen anyway, why not remove `std::strnlen` completely for now? That would reduces conditional compilation and I don't believe this will cause a significant performance problem.

Will c11's strnlen_s be part of c++17?

**Comment by rianquinn** [comment]
@MikeGitb Correct me if I'm wrong, but I think the `cstring` headers are just namespaced wrappers around existing C functions, and thus `std::strnlen` doesn't exist either. If that's not the case then I would agree that it would be the better approach. 

@Josaphat We had the same issue with our cross compiler for [Bareflank](https://github.com/Bareflank/hypervisor). If you created the cross compiler yourself, you can add `-DNO_IMPLICIT_EXTERN_C=1` while compiling GCC and it will fix that issue so that you can use `-isystem`. Either way, I think the implicit narrowing is due to the fact that span uses a signed int and not an unsigned int, so either way, I think a PR should be put in place to address that issue, and the narrowing should not explicit.

**Comment by Josaphat** [comment]
@MikeGitb If the consensus is that removing the use of `strnlen` entirely is the way to go, I'm happy to do that. And I don't know whether C++17 is meant to switch to C11 instead of C99.

@rianquinn We don't currently compile the toolchain ourselves but we would be open to doing so. However, I think it should compile cleanly in all cases.

With respect to the narrowing, I think it's safe enough if it's checked, and your program probably has a bug if you try to create a string larger than `INT_MAX`. Some coding standards eschew the use of unsigned integral types in favor of signed types everywhere (I haven't been subject to such a coding standard) so it's not unreasonable to make the conversion.

**Comment by rianquinn** [comment]
Agreed, the narrow should be added.

**Comment by MikeGitb** [comment]
@rianquinn: Yes, sorry, that was a typo / unclear expression on my side - what I meant was to not use the system/compiler provided `strnlen` at all, but only used the manual implementation of the `string_length` function provided in #334. 

c++17 is meant to switch to C11, but I'm not 100% sure what that means exactly.

Regarding the narrowing: shouldn't `gsl::detail::string_length` just take std::ptrdiff_t as an argument, as this is used as a size type throughout the gsl anyway? Otherwise we are first narrowing from `std::ptrdiff_t` to `size_t` at the callsite and then again from `std::size_t` to `std::ptrdiff_t` in the function body.

**Comment by Josaphat** [comment]
You're right. Every time `details::string_length` is used, the second parameter is cast from `std::ptrdiff_t` to `size_t`, and the result is cast from `size_t` to `std::ptrdiff_t`. I'll switch them all to use `ptrdiff_t` and remove the casts.

**Comment by neilmacintosh** [comment]
@Josaphat @MikeGitb that would be a nice improvement to make...

**Comment by neilmacintosh** [comment]
This has been fixed by #391 - thanks everyone!
